### PR TITLE
Update no_trainer.py scripts to include accelerate gradient accumulation wrapper

### LINKED ISSUE #18436: Update no_trainer scripts to include gradient accumulation
### Feature request

🤗 Accelerate has a gradient accumulation wrapper, and the `no_trainer` scripts should be updated to include it! 

An example can be seen [here](https://github.com/huggingface/accelerate/blob/main/examples/by_feature/gradient_accumulation.py), below is an example diff of what the integration would look like:

```diff
-     accelerator = (
-         Accelerator(log_with=args.report_to, logging_dir=args.output_dir) if args.with_tracking else Accelerator()
-     )
+     accelerator = (
+         Accelerator(log_with=args.report_to, logging_dir=args.output_dir, gradient_accumulation_steps=args.gradient_accumulation_steps) if args.with_tracking else Accelerator(gradient_accumulation_steps=args.gradient_accumulation_steps)
+     )
```

As well as:
```diff
for step, batch in enumerate(train_dataloader):
+     with accelerator.accumulate(model):
```
```diff
-             loss = loss / args.gradient_accumulation_steps
            accelerator.backward(loss)
-             if step % args.gradient_accumulation_steps == 0 or step == len(train_dataloader) - 1:
            optimizer.step()
            lr_scheduler.step()
            optimizer.zero_grad()
                progress_bar.update(1)
                completed_steps += 1
```

### Motivation

This is a great first issue for someone who wants to learn how to use some of the latest bits in Accelerate and get an easy beginner contribution to the library 🤗 

### Your contribution

If you decide to pick up this issue, feel free to ping myself, @sgugger, or @sourabh112 to review 🤗

### PR DESCRIPTION
**Pull request by Rasmusafj**
# What does this PR do?
Updates no_trainer.py scripts to use the new gradient accumulation wrapper feature from accelerate according to #18436. 

## Before submitting
- [ N] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ Y] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ N] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [N ] Did you write any new necessary tests?

## Who can review?
@sgugger

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: examples/pytorch/image-classification/run_image_classification_no_trainer.py
Comment (sgugger) [inline examples/pytorch/image-classification/run_image_classification_no_trainer.py:222]: Maybe define `log_kwargs` in an if/else then do just one call to `Accelerator`?

File: examples/pytorch/image-classification/run_image_classification_no_trainer.py
Comment (sgugger) [inline examples/pytorch/image-classification/run_image_classification_no_trainer.py:394]:
```suggestion
        num_training_steps=len(train_dataloader) * args.num_train_epochs,
```
This length should be an integer.

Also we are ignoring the `max_steps` argument set by the user here, I don't think this change should be made.

File: examples/pytorch/image-classification/run_image_classification_no_trainer.py
Comment (sgugger) [inline examples/pytorch/image-classification/run_image_classification_no_trainer.py:489]: I think it's going to be easier to just always update, at least for now.

**Review by sgugger — COMMENTED** [review]
I think we need also change the definition of `num_update_steps_per_epoch` at line 385 and also multiply `max_steps` by the number of gradient accumulation steps, as discussed on the issue.

Thanks a lot for your PR!

File: examples/pytorch/image-classification/run_image_classification_no_trainer.py
Comment (Rasmusafj) [reply examples/pytorch/image-classification/run_image_classification_no_trainer.py:489]: If we always update and we have gradient_accumulation_steps > 1, the progress bar will act wrong i.e. increment fast and then stop while still performing updates.

This is because if max_train_steps is not given as an argument, its calculated based on dividing with gradient_accumulation_steps.

File: examples/pytorch/image-classification/run_image_classification_no_trainer.py
Comment (Rasmusafj) [reply examples/pytorch/image-classification/run_image_classification_no_trainer.py]: I tried doing what you describe. Let me know if this is correct. :)

File: examples/pytorch/image-classification/run_image_classification_no_trainer.py
Comment (Rasmusafj) [reply examples/pytorch/image-classification/run_image_classification_no_trainer.py]: Yea, I see. 

So, now i've added that its `len(train_dataloader) * args.num_train_epochs` and `args.max_train_steps * args.gradient_accumulation_steps` in case user provided max_train_steps.

File: examples/pytorch/image-classification/run_image_classification_no_trainer.py
Comment (muellerzr) [reply examples/pytorch/image-classification/run_image_classification_no_trainer.py:490]: Excellent point. I think the proper solution here would be a small wrapper around the progress bar (on top of the new one Accelerate just added) that is tied to the `GradientState` similar to how `optimizer.step()` works, where if the gradients should be synchronized then we assume a step has been performed and the progress bar moves. Otherwise it does nothing. 

But for now what you have here does look right to me.

**Review by sgugger — APPROVED** [review]
LGTM, let me know if you want to do more examples in the same PR or if this is good to merge!

**Comment by sgugger** [comment]
Let us know when the PR is ready for review (it's in draft mode right now) so we can go ahead and merge.

**Comment by Rasmusafj** [comment]
@sgugger  @muellerzr I changed PR to review. :)

File: examples/pytorch/question-answering/run_qa_beam_search_no_trainer.py
Comment (sgugger) [inline examples/pytorch/question-answering/run_qa_beam_search_no_trainer.py:751]: This should be `args.gradient_accumulation_steps`  here.

File: examples/pytorch/speech-pretraining/run_wav2vec2_pretraining_no_trainer.py
Comment (sgugger) [inline examples/pytorch/speech-pretraining/run_wav2vec2_pretraining_no_trainer.py:371]: This one does a lot more things in its training loop, so would leave it as it was as I'm more afraid to break anything.

**Review by sgugger — COMMENTED** [review]
Thanks a lot, I wouldn't touch the Wav2Vec2 pretraining script as it is very different from the others and I'd need the speech team to validate those changes (you can make another PR with just it, but I think it's also fine to just leave it as is).

Apart from a typo, all the other ones look good to me!

File: examples/pytorch/question-answering/run_qa_beam_search_no_trainer.py
Comment (Rasmusafj) [reply examples/pytorch/question-answering/run_qa_beam_search_no_trainer.py]: fixed. :)

File: examples/pytorch/speech-pretraining/run_wav2vec2_pretraining_no_trainer.py
Comment (Rasmusafj) [reply examples/pytorch/speech-pretraining/run_wav2vec2_pretraining_no_trainer.py]: Removed changes

**Comment by Rasmusafj** [comment]
@sgugger I removed changes to wav2vec script and fixed the using of wrong constant. 

Feel sure to merge if its fine. Its going to be squash merged right? or do i need to rebase and squash before merge?

**Comment by sgugger** [comment]
We squash indeed. Thanks again for your contribution!

**Pull request merged**
